### PR TITLE
Adding GC Stats to TaskMetrics (and three small fixes)

### PR DESCRIPTION
**Pull request by pwendell**
The main change here is that we add time the JVM spent in GC for reach task to `TaskMetrics` and the web UI.

There are also three minor fixes related to terminology/rendering of the UI pages.

### CODE REVIEW

File: core/src/main/scala/spark/scheduler/local/LocalScheduler.scala [hunk 1]
```diff
@@ -210,7 +218,10 @@ private[spark] class LocalScheduler(threads: Int, val maxFailures: Int, val sc:
       case t: Throwable => {
         val serviceTime = System.currentTimeMillis() - taskStart
         val metrics = attemptedTask.flatMap(t => t.metrics)
-        metrics.foreach{m => m.executorRunTime = serviceTime.toInt}
+        metrics.foreach{ m =>
```
Comment (mateiz) [inline core/src/main/scala/spark/scheduler/local/LocalScheduler.scala]: Add a space before the {

File: core/src/main/scala/spark/executor/Executor.scala [hunk 2]
```diff
@@ -155,7 +158,10 @@ private[spark] class Executor(executorId: String, slaveHostname: String, propert
         case t: Throwable => {
           val serviceTime = (System.currentTimeMillis() - taskStart).toInt
           val metrics = attemptedTask.flatMap(t => t.metrics)
-          metrics.foreach{m => m.executorRunTime = serviceTime}
+          metrics.foreach{m =>
```
Comment (mateiz) [inline core/src/main/scala/spark/executor/Executor.scala]: Add a space before the {

File: core/src/main/scala/spark/scheduler/local/LocalScheduler.scala [hunk 3]
```diff
@@ -202,6 +209,7 @@ private[spark] class LocalScheduler(threads: Int, val maxFailures: Int, val sc:
       val serviceTime = System.currentTimeMillis() - taskStart
       logInfo("Finished " + taskId)
       deserializedTask.metrics.get.executorRunTime = serviceTime.toInt
+      deserializedTask.metrics.get.jvmGCTime = getTotalGCTime - startGCTime
```
Comment (rxin) [inline core/src/main/scala/spark/scheduler/local/LocalScheduler.scala:211]: Does this lead to double counting if two tasks are running at the same time and a GC happens?

Comment (rxin) [inline core/src/main/scala/spark/scheduler/local/LocalScheduler.scala:211]: It is actually accurate at a per-task level - we just can't report the aggregated time. Do we ever aggregate them?

Comment (pwendell) [inline core/src/main/scala/spark/scheduler/local/LocalScheduler.scala:211]: GC stops the entire JVM, so it does "double count" but these are IMO correct. The main point here is to make it easy to see if a slow task is caused by a GC pause.

We can add total time spent in GC as well in a subsequent PR. It's a bit confusing how you should count the total though... for instance do you "double count" or not, if multiple tasks are running during a GC event. Another useful thing to add would be at the executor level add a metric for what proportion of overall time is spent in GC.

Comment (rxin) [inline core/src/main/scala/spark/scheduler/local/LocalScheduler.scala:211]: For young gen concurrent gc's, are the time reported here too?

I think this is ok as long as we don't report aggregated gc time based on these numbers, for reasons pointed out by both of us earlier.

Comment (pwendell) [inline core/src/main/scala/spark/scheduler/local/LocalScheduler.scala:211]: It's up the the individual collector to decide what to report. I'm not sure how the young collector in marksweep (called `PS Scavenge`) returns collection time.

Here is an example of code that treats collectors differently, but they just do young vs old (whereas we probably want concurrent vs not).
http://svn.cometd.org/trunk/cometd-java/cometd-java-client/src/test/java/org/cometd/client/StatisticsHelper.java

Comment (pwendell) [inline core/src/main/scala/spark/scheduler/local/LocalScheduler.scala:211]: One thing is we could exclude `PS Scavenge` if it's present... do you think that would be better? Then we'd only include stop-the-world events:

Also this is a good link as well:
http://docs.neo4j.org/chunked/stable/configuration-jvm.html

Comment (pwendell) [inline core/src/main/scala/spark/scheduler/local/LocalScheduler.scala:211]: Actually @rxin - maybe we should stick with the simplest thing at first, just use the JVM's own statistic. We can do finer grained reporting a bit later, such as breaking down between old/new GC and concurrent/stop the world.

Comment (rxin) [inline core/src/main/scala/spark/scheduler/local/LocalScheduler.scala:211]: i agree

**Comment by pwendell** [comment]
Okay I addressed @mateiz's comments and looks like @rxin is okay with this as-is. Let me know if there are any other comments.

File: core/src/main/scala/spark/ui/jobs/StagePage.scala [hunk 4]
```diff
@@ -154,10 +157,17 @@ private[spark] class StagePage(parent: JobProgressUI) {
...
+      {if (shuffleRead) {
+        <td>{metrics.flatMap{m => m.shuffleReadMetrics}.map{s =>
+          Utils.memoryBytesToString(s.remoteBytesRead)}.getOrElse("")}</td>
+      }}
+      {if (shuffleWrite) {
+        <td>{metrics.flatMap{m => m.shuffleWriteMetrics}.map{s =>
+          Utils.memoryBytesToString(s.shuffleBytesWritten)}.getOrElse("")}</td>
+      }}
+      <td sorttable_customkey={gcTime.toString}>
+        {if (gcTime > 0) {parent.formatDuration(gcTime)} else ""}
```
Comment (mateiz) [inline core/src/main/scala/spark/ui/jobs/StagePage.scala]: Does this need braces around parent.formatDuration?

File: core/src/main/scala/spark/scheduler/local/LocalScheduler.scala [hunk 5]
```diff
@@ -31,6 +34,7 @@ import spark.scheduler._
 import spark.scheduler.cluster._
 import spark.scheduler.cluster.SchedulingMode.SchedulingMode
 import akka.actor._
+import management.ManagementFactory
```
Comment (mateiz) [inline core/src/main/scala/spark/scheduler/local/LocalScheduler.scala]: Use the full package name for this (java.lang.management.ManagementFactory)

File: core/src/main/scala/spark/executor/Executor.scala [hunk 6]
```diff
@@ -155,7 +158,10 @@ private[spark] class Executor(executorId: String, slaveHostname: String, propert
         case t: Throwable => {
           val serviceTime = (System.currentTimeMillis() - taskStart).toInt
           val metrics = attemptedTask.flatMap(t => t.metrics)
-          metrics.foreach{m => m.executorRunTime = serviceTime}
+          metrics.foreach {m =>
```
Comment (mateiz) [inline core/src/main/scala/spark/executor/Executor.scala]: For these, I actually meant the right style is `metrics.foreach { m =>` when they're multiline, but on second thought, maybe just do `for (m <- metrics)`

File: core/src/main/scala/spark/scheduler/local/LocalScheduler.scala [hunk 7]
```diff
@@ -210,7 +218,10 @@ private[spark] class LocalScheduler(threads: Int, val maxFailures: Int, val sc:
       case t: Throwable => {
         val serviceTime = System.currentTimeMillis() - taskStart
         val metrics = attemptedTask.flatMap(t => t.metrics)
-        metrics.foreach{m => m.executorRunTime = serviceTime.toInt}
+        metrics.foreach {m =>
```
Comment (mateiz) [inline core/src/main/scala/spark/scheduler/local/LocalScheduler.scala]: Ditto here

File: core/src/main/scala/spark/scheduler/local/LocalScheduler.scala (hunk 5 above)
Comment (mateiz) [inline core/src/main/scala/spark/scheduler/local/LocalScheduler.scala]: Actually this looks like a duplicate import anyway

**Comment by mateiz** [comment]
Just saw some other minor issues, but otherwise it seems good.

File: core/src/main/scala/spark/executor/Executor.scala (hunk 6 above)
Comment (pwendell) [inline core/src/main/scala/spark/executor/Executor.scala]: Where in the style guide does it say to do that? I could only find a single-line reference for foreach and it didn't put a space after the open paren:
http://docs.scala-lang.org/style/method-invocation.html

Comment (mateiz) [inline core/src/main/scala/spark/executor/Executor.scala]: I'm just saying based on other code we have that uses braces. For parens it doesn't make sense.

**Comment by pwendell** [comment]
@mateiz thanks for the style catches. These are fixed now.

**Comment by mateiz** [comment]
Cool, thanks.

**Pull request merged**